src/validation.cpp

```diff
@@ -2603,6 +2608,180 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
+                        if (!res)
+                            LogPrintf("Can't refund makerDeposit back to owner: %s\n", res.msg);
+                        refund = true;
+                    }
+                }
+                else if (status == CICXSubmitDFCHTLC::STATUS_REFUNDED)
+                    refund = true;
+
+                if (refund)
+                {
+                    CTokenAmount amount;
+                    CScript ownerAddress;
+                    amount = {order->idToken, dfchtlc->amount};
+                    if (order->orderType == CICXOrder::TYPE_INTERNAL)
+                        ownerAddress = order->ownerAddress;
+                    else if (order->orderType == CICXOrder::TYPE_EXTERNAL)
+                        ownerAddress = offer->ownerAddress;
+
+                    CScript txidaddr = CScript(dfchtlc->creationTx.begin(),dfchtlc->creationTx.end());
+                    auto res = cache.SubBalance(txidaddr,amount);
+                    if (!res)
+                        LogPrintf("Can't subtract balance from dfc htlc txidaddr: %s\n", res.msg);
+                    if (res.ok)
+                    {
+                        res = cache.AddBalance(ownerAddress,amount);
+                        if (!res)
+                            LogPrintf("Can't add balance from dfc htlc back to owner: %s\n", res.msg);
+                    }
+                    if (res.ok)
+                    {
+                        res = cache.ICXCloseDFCHTLC(*dfchtlc,status);
+                        if (!res)
+                            LogPrintf("Can't refund dfc htlc: %s\n", res.msg);
+                    }
+                }
+
+                return true;
+            },pindex->nHeight);
+
+            cache.ForEachICXSubmitEXTHTLCExpire([&](CICXOrderView::StatusKey const & key, uint8_t status) {
+                if (static_cast<int>(key.first) != pindex->nHeight) return (false);
+
+                auto exthtlc = cache.GetICXSubmitEXTHTLCByCreationTx(key.second);
+                if (!exthtlc)
+                    return true;
+                auto offer = cache.GetICXMakeOfferByCreationTx(exthtlc->offerTx);
+                if (!offer)
+                    return true;
+                auto order = cache.GetICXOrderByCreationTx(offer->orderTx);
+                if (!order)
+                    return true;
+
+                if (status == CICXSubmitEXTHTLC::STATUS_EXPIRED && order->orderType == CICXOrder::TYPE_INTERNAL)
+                {
+                    std::unique_ptr<CICXSubmitDFCHTLCImplemetation> dfchtlc;
+                    cache.ForEachICXSubmitDFCHTLCOpen([&](CICXOrderView::TxidPairKey const & key, uint8_t i) {
+                        if (key.first == exthtlc->offerTx)
+                        {
+                            dfchtlc = pcustomcsview->GetICXSubmitDFCHTLCByCreationTx(key.second);
+                            return (false);
+                        }
+                        return false;
```

Comment:
This lambda function always return false, so only search one item?

Reply:
Yes, this is intended behavior. It should check if DFC or in below cased EXT HTLC exists. So if it doesn't exists it returns maker deposit to the maker because taker didn't create his htlc in period of 100 DFC blocks.